### PR TITLE
Use YAML for route-sync config

### PR DESCRIPTION
**Pull request by johnsonj**
- Update the config package to load and verify a `route-sync.yml` file for configuration. 
- Perform deserialization via `Schema` objects, inspired by the [route-registrar](https://github.com/cloudfoundry/route-registrar/blob/master/config/config.go#L27)
- Update a test bug where the table based tests captures the last value of the list. [more info](https://github.com/pivotal-cf-experimental/kubo-release/compare/PR-route-sync-use-yaml-for-config?expand=1#diff-ceecf21dea42bd17522a7c945a5b4a2aR42). Semi-unrelated but I added the first usage of the DescribeTable extension in this PR.
- Update BOSH job to use YAML config
- Update developer docs/config example for YAML

Notes:
- Validation in `ConfigSchema.ToConfig()` could benefit from some reflection (lots of len(..) == ..). I decided this would probably obfuscate things too much given this code won't need to change often. Open to discussion.

### CODE REVIEW

File: src/route-sync/config/config.go [hunk 1]
```diff
@@ -1,47 +1,148 @@
...
 	cfConfig "code.cloudfoundry.org/route-registrar/config"
-	"errors"
-	"github.com/kelseyhightower/envconfig"
 	uaaconfig "code.cloudfoundry.org/uaa-go-client/config"
 )
 
-// Config contains settings for the route-sync with envconfig annotations
+// ConfigSchema defines the file schema for the YAML configuration of route-sync
+type ConfigSchema struct {
+	NatsServers               []MessageBusServerSchema `yaml:"nats_servers"`
```
Comment (jhvhs) [inline src/route-sync/config/config.go]: The `cfConfig.MessageBusServer` is able to map directly without the extra YAML tags

File: src/route-sync/main.go [hunk 2]
```diff
@@ -31,7 +32,19 @@ func main() {
 }
 
 func loadConfig(logger lager.Logger) *config.Config {
-	cfg, err := config.NewConfig()
+	var configPath string
+
+	flags := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	flags.StringVar(&configPath, "configPath", "", "path to configuration file with json encoded content")
+	flags.Set("configPath", "route-sync.yml")
```
Comment (jhvhs) [inline src/route-sync/main.go]: Why aren't we using `#StringVar` to set the default value?

File: src/route-sync/main.go [hunk 3]
```diff
@@ -31,7 +32,19 @@ func main() {
 }
 
 func loadConfig(logger lager.Logger) *config.Config {
-	cfg, err := config.NewConfig()
+	var configPath string
+
+	flags := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
```
Comment (jhvhs) [inline src/route-sync/main.go]: Why do we have to use the `FlagSet` directly, as opposed to `flag.*Var`?

File: src/route-sync/application/application_test.go [hunk 4]
```diff
@@ -24,9 +25,8 @@ var _ = Describe("Application", func() {
 		sourceFake *routefakes.FakeSource
 		routerFake *routefakes.FakeRouter
 		cfg        = &config.Config{
-			RawNatsServers:            "[{\"Host\": \"host\",\"User\": \"user\", \"Password\": \"password\"}]",
 			NatsServers:               []cfConfig.MessageBusServer{{Host: "host", User: "user", Password: "password"}},
-			RoutingApiUrl:             "https://api.cf.example.org",
+			RoutingAPIURL:             "https://api.cf.example.org",
```
Comment (johnsonj) [inline src/route-sync/application/application_test.go]: Any objection to removing all of the upper-casing here? Any rule on when/when not to case makes this struct pretty inconsistent.

File: src/route-sync/config/config.go (hunk 1 above)
Comment (johnsonj) [inline src/route-sync/config/config.go]: This let's us do validation the same way as we do for ConfigSchema->Config

File: jobs/route-sync/templates/config/route-sync.yml.erb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+message_bus_servers = nats_ips.map do |ip|
+{
+ host: "#{ip}:#{nats_port}",
+ user: nats_user,
+ password: nats_password
+}
+end
+%>
+
+nats_servers: <%=JSON.pretty_generate(message_bus_servers) %>
```
Comment (johnsonj) [inline jobs/route-sync/templates/config/route-sync.yml.erb]: constructing the object manually per recommendation

File: src/route-sync/config/config.go [hunk 6]
```diff
@@ -1,47 +1,148 @@
...
+	"code.cloudfoundry.org/multierror"
 	cfConfig "code.cloudfoundry.org/route-registrar/config"
-	"errors"
-	"github.com/kelseyhightower/envconfig"
 	uaaconfig "code.cloudfoundry.org/uaa-go-client/config"
 )
 
-// Config contains settings for the route-sync with envconfig annotations
+// ConfigSchema defines the file schema for the YAML configuration of route-sync
+type ConfigSchema struct {
```
Comment (johnsonj) [inline src/route-sync/config/config.go]: See PR description. I got this pattern from `route-registrar/config`. 

The good:

- We will not produce invalid `Config` objects. We refuse to convert if the ConfigScheme isn't valid. No awkward partially constructed object problems.
- Persistence is clearly separated from runtime config. These are two different concerns and makes testing them separately very straightforward. 

The bad:
- Duplication

File: src/route-sync/main.go (hunk 3 above)
Comment (johnsonj) [inline src/route-sync/main.go]: Lack of experience with flags

File: src/route-sync/main.go (hunk 2 above)
Comment (johnsonj) [inline src/route-sync/main.go]: (same comment as above, addressing)

File: src/route-sync/config/config_test.go [hunk 7]
```diff
@@ -1,79 +1,155 @@
...
+	"github.com/onsi/gomega/gbytes"
+
+	cfConfig "code.cloudfoundry.org/route-registrar/config"
+	uaaconfig "code.cloudfoundry.org/uaa-go-client/config"
 )
 
 var _ = Describe("Config", func() {
-	Context("Configuration is set", func() {
+	Context("Reading schema from a file that doesn't exist", func() {
+		It("does not return a config schema", func() {
```
Comment (alex-slynko) [inline src/route-sync/config/config_test.go]: I think "returns error" reads better

File: src/route-sync/config/config_test.go [hunk 8]
```diff
@@ -1,79 +1,155 @@
...
-			c, err := NewConfig()
-
-			Expect(err).NotTo(HaveOccurred())
-			Expect(c.NatsServers).To(HaveLen(1))
-
-			natsServer := c.NatsServers[0]
-			Expect(natsServer.Host).To(Equal("10.0.1.8:4222"))
-			Expect(natsServer.User).To(Equal("nats"))
-			Expect(natsServer.Password).To(Equal("natspass"))
+		Context("with an empty file", func() {
```
Comment (alex-slynko) [inline src/route-sync/config/config_test.go]: I don't think we need two test cases here.
If file is there and proper formatted, we should expect `ConfigSchema` to be returns.

File: src/route-sync/config/config.go (hunk 6 above)
Comment (alex-slynko) [inline src/route-sync/config/config.go]: ok.

File: src/route-sync/config/config_test.go (hunk 7 above)
Comment (johnsonj) [inline src/route-sync/config/config_test.go]: agreed

File: src/route-sync/config/config_test.go (hunk 8 above)
Comment (johnsonj) [inline src/route-sync/config/config_test.go]: good point, this is redundant

File: src/route-sync/glide.yaml [hunk 9]
```diff
@@ -15,10 +15,12 @@ import:
 - package: github.com/kelseyhightower/envconfig
```
Comment (alex-slynko) [inline src/route-sync/glide.yaml]: one last thing. We don't need `envconfig` anymore

**Pull request merged**